draft-pda-core.md

```diff
@@ -291,11 +291,437 @@ Each PA task in this document is divided into three sub-protocols as follows.
+  select (PAClientParam.task.proto) {
+    case prio: PrioParam; // Defined in Section 3
+    case hits: HitsParam; // Defined in Section 4
+  }
+} PAParam;
+~~~
+
+[TODO: Add batch parameters, including min/max batch size, batch time window,
+etc.]
+
+
+### Helper key configuration
+
+Each helper specifies the
+[HPKE](https://datatracker.ietf.org/doc/draft-irtf-cfrg-hpke/) public key the
+client will use to encrypt the helper's share of the report. The public key and
+associated parameters are structured as follows:
+
+~~~
+struct {
+  uint8 id;
+  HpkeKemId kem_id;
+  HpkeKdfId kdf_id;
+  HpkeAeadKdfId aead_id;
+  HpkePublicKey public_key;
+} HPKEKeyConfig;
+
+opaque HpkePublicKey<0..2^16-1>;
+uint16 HpkeAeadId; // Defined in I-D.irtf-cfrg-hpke
+uint16 HpkeKemId;  // Defined in I-D.irtf-cfrg-hpke
+uint16 HpkeKdfId;  // Defined in I-D.irtf-cfrg-hpke
+~~~
+
+We call this the helper's *key configation*. The key configuration is used to
+set up a base-mode HPKE context to use to derive symmetric keys for protecting
+the shares sent to the helper. The *config id*, `HPKEKeyConfig.id`, is forwarded
+by the client to the helper, who uses this value to decide if it knows how to
+decrypt a share it receives.
+
+**Pre-conditions.**
+We assume the following conditions hold before the protocol begins:
+1. The client, aggregators, and collector are configured with a specific PA task.
+1. The client knows the URL of the leader endpoint, e.g., `example.com/metrics`.
+   We write this URL as `[leader]` below. (We write `[helper]` for a helper's
+   URL.)
+1. The client and leader can establish a leader-authenticated TLS channel.
+1. The leader and each helper can establish a leader-authenticated TLS channel.
+1. Each helper has chosen an HPKE key pair.
+1. The aggregators agree on a set of PA tasks, as well as the PA protocol and
+   parameters used for each task.
+
+## Upload {#pa-upload}
+
+[TODO: Add an illustration of this sub-protocol.]
+
+Uploading a report involves two requests to the leader. In the  *upload start
+request*, the client discovers the protocol-specific parameters and the endpoint
+URL of each helper. In the *upload finish request* uploads its report to the
+leader.
+
+[OPEN ISSUE: @acmiyaguchi pointed out that the use of an anonymizing proxy for
+uploading shares might be easier to implement if the "upload" phase involved a
+single HTTP request. Is this true? Does OHTTP
+(https//www.ietf.org/archive/id/draft-thomson-http-oblivious-01.html) allow
+clients to make multiple requests without too much fuss?]
+
+### Upload Start
+
+**Request.**
+Let `[leader]` denote the URL of the leader's report ingestion endpoint. The
+client sends a POST request to `[leader]/upload_start` with the following
+message:
+
+~~~
+struct {
+  PATask task;
+  opaque id[16];
+} PAUploadStartReq;
+~~~
+
+The first field, `task` corresponds to the PA task for which a report will be
+generated. The second, `id` is a random, 16-byte string generated by the
```

Comment:
The `id` seems to tie the upload start and finish together. A couple questions:
1. Can we drop it if we turn this into a single HTTP request?
2. if we keep both requests, why do they need to be tied together?

Reply:
>  The id seems to tie the upload start and finish together. A couple questions:
>
>     1. Can we drop it if we turn this into a single HTTP request?

For some Prio proof systems, it's useful for the leader to send the client a "challenge" that it'll use to generate the proof. One way to use the id is as a way of generating the challenge in a way that's unique for each upload.

>     2. if we keep both requests, why do they need to be tied together?

Retransmitting the id allows the two HTTP requests to be handled without carrying state between them.

Reply:
I don't see why that's valuable or desired, since it just pushes state onto the server. (Left a separate comment for that, so will resolve this.)